src/os/spawn.php

```diff
@@ -0,0 +1,173 @@
+<?hh
+
+namespace HH\Lib\Experimental\OS;
+use namespace HH\Lib\C;
+use namespace HH\Lib\Vec;
+use namespace HH\Lib\_Private\_OS;
+use type HH\Lib\OS\{FileDescriptor, IsNotADirectoryException};
+use function HH\Lib\OS\{close, open};
+
+newtype pid_t as int = int;
+newtype ExitCode as int = int;
+
+<<__Sealed(
+  posix_spawn_file_actions_addopen::class,
+  posix_spawn_file_actions_adddup2::class,
+  posix_spawn_file_actions_addchdir_np::class,
+)>>
+interface PosixSpawnFileActionsSetter {}
+
+final class posix_spawn_file_actions_addopen
+  implements PosixSpawnFileActionsSetter {
+  public function __construct(
+    public int $filedes,
+    public string $path,
+    public int $oflag,
+    public int $mode,
+  ) {}
+}
+
+final class posix_spawn_file_actions_adddup2
+  implements PosixSpawnFileActionsSetter {
+  public function __construct(
+    public FileDescriptor $filedes,
+    public int $newfiledes,
+  ) {}
+}
+
+final class posix_spawn_file_actions_addchdir_np
+  implements PosixSpawnFileActionsSetter {
+  public function __construct(public string $path) {
+  }
+}
+
+type posix_spawn_file_actions_t = vec<PosixSpawnFileActionsSetter>;
+
+const int POSIX_SPAWN_SETPGROUP = 2;
+const int POSIX_SPAWN_SETSID = 128;
+
+type PosixSpawnFlags = int;
+
+type posix_spawnattr_t = shape(
+  ?'posix_spawnattr_setpgroup' => pid_t,
+  ?'posix_spawnattr_setflags' => PosixSpawnFlags,
+);
+
+function posix_spawnp(
+  string $file,
+  posix_spawn_file_actions_t $file_actions,
+  posix_spawnattr_t $attributes,
+  vec<string> $argv,
+  vec<string> $envp,
+): pid_t {
+
+  // $default_fork_and_execve will call _OS\fork_and_execve but redirect
+  // standard I/O to /dev/null by default.
+  $default_fork_and_execve = (
+    string $path,
+    vec<string> $argv,
+    vec<string> $envp,
+    dict<int, FileDescriptor> $fds,
+
+    // Use inline shape type instead of _OS\ForkAndExecveOptions as a
+    // workaround to https://github.com/facebook/hhvm/issues/8989
+    shape(
+      ?'cwd' => string,
+      ?'setsid' => bool,
+      ?'execvpe' => bool,
+      ?'setpgid' => int,
+    ) $options,
+  ) ==> {
+    $cwd = Shapes::idx($options, 'cwd');
+    $dev_null_fd = open('/dev/null', _OS\O_RDWR);
+    try {
+      $fds[_OS\STDIN_FILENO] ??= $dev_null_fd;
+      $fds[_OS\STDOUT_FILENO] ??= $dev_null_fd;
+      $fds[_OS\STDERR_FILENO] ??= $dev_null_fd;
+      return _OS\fork_and_execve(
+        $cwd is null ? $path : $cwd.'/'.$path,
+        $argv,
+        $envp,
+        $fds,
+        $options,
+      );
+    } finally {
+      close($dev_null_fd);
+    }
+  };
+
+  // Each file action is interpreted as an wrapper of previous $fork_and_execve
+  // function, adding more options and flags.
```

Comment:
This is leading to deep nesting and a lot of boilerplate (for definitions); is this a style preference, or are there behavior benefits over modifying state?

Reply:
To interpret `posix_spawn_file_actions_addopen`, we must inject an additional `try` `finally` block to close the file descriptor created. Wrapping previous `$fork_and_execve` would be an easy to implement the `try` / `finally` block.

Reply:
if it simplifies the code significantly, we could just not support _addopen (for now?) - and it's not necessary - as we'll be openining it within hack code anyway

Reply:
I don't think this approach adds significant more boilerplate than other approaches. Any way we need some states when iterating through the file actions. If we don't put the states on parameters, we will put states on a `tuple` or a `shape`, which, in my opinion, are the same complex as the states on parameters.